Source/UI/NewUI/BillingForm/View/Field/BillingFormPhoneNumberText.swift

```diff
@@ -0,0 +1,76 @@
+import UIKit
+import PhoneNumberKit
+
+final class BillingFormPhoneNumberText: BillingFormTextField {
+
+    var phoneNumber: PhoneNumber?
+    var nationalNumber: String {
+        let rawNumber = self.text ?? ""
+        return partialFormatter.nationalNumber(from: rawNumber)
+    }
+    var isValidNumber: Bool {
+        let rawNumber = self.text ?? ""
+        do {
+            phoneNumber = try phoneNumberKit.parse(rawNumber.replacingOccurrences(of: " ", with: ""))
+            return true
+        } catch {
+            return false
+        }
+    }
+    
+    private var previousTextCount = 0
+    private var previousFormat = ""
+    private let phoneNumberKit = PhoneNumberKit()
+    private lazy var partialFormatter = PartialFormatter(phoneNumberKit: phoneNumberKit, defaultRegion: "GB", withPrefix: true)
+
+    override init(type: BillingFormCellType, tag: Int) {
+        super.init(type: type,tag: tag)
+        setup()
+    }
+    
+    required init?(coder: NSCoder) {
+        fatalError("init(coder:) has not been implemented")
+    }
+    
+    private func setup() {
+        keyboardType = .phonePad
+        addTarget(self, action: #selector(textFieldDidChange), for: UIControl.Event.editingChanged)
+    }
+
+    /// Called when the text changed.
+    @objc private func textFieldDidChange(textField: UITextField) {
+        var targetCursorPosition = 0
+        if let startPosition = textField.selectedTextRange?.start {
+            targetCursorPosition = textField.offset(from: textField.beginningOfDocument, to: startPosition)
+        }
+
+        let phoneNumber = textField.text!
+        let formatted = partialFormatter.formatPartial(phoneNumber)
+        textField.text = formatted
+
+        if var targetPosition = textField.position(from: textField.beginningOfDocument, offset: targetCursorPosition) {
+            if targetCursorPosition != 0 {
+                let lastChar = formatted[formatted.index(formatted.startIndex,
+                                                         offsetBy: targetCursorPosition - 1)]
+                if lastChar == " " && previousTextCount < formatted.count && phoneNumber != formatted {
```

Comment:
suggestion : You can create a let constant as string called whitespace and use it elsewhere instead of magic white space 
let whiteSpace = " "

Reply:
This part is from old code, it will be refactored next country ticket.